### PR TITLE
fix(messaging)!: Fix mix up messaging in window and custom event messengers

### LINKED ISSUE #57: @webext-core/messaging errors with No response when sending data back from an injected script
Using only example code from the [webext-core site](https://webext-core.aklinker1.io/guide/messaging/), injected script messaging causes an error (No error when sending messages *to* the injected script)
Potentially related to #55 ?

Error: 
```
Uncaught (in promise) Error: No response
    at Object.<anonymous> (chunk-ALTBYIMW.js:80:67)
    at Generator.next (<anonymous>)
    at fulfilled (chunk-ALTBYIMW.js:36:24)
```

Minimal repro: https://github.com/rcoopr/wxtcore-messaging-repro
  - clone
  - `pnpm i`
  - `pnpm dev`
  - check console

What steps did I do to create example:
  - init WXT using `pnpm dlx wxt@latest init <project-name>`
  - `pnpm i @webext-core/messaging`
  - Add permissions for injected script to wxt config
  - Copy code from https://webext-core.aklinker1.io/guide/messaging/#window-messaging (Also, there is a typo, `browser.runtime.getUrl` is actually `getURL`)
  - run `pnpm dev` and check console

The messages are recieved, yet there is an error anyway. The issue is also present in the built output. Only tested on Chrome 122

### PR DESCRIPTION
**Pull request by 1natsu172**
fix #57 

## Firstly

It may be hard to review, so take your time. 

Also, if approved, it might be safer to make it a major bump. Because the messaging path is narrowed and pseudo-channeled, it is possible that a user's application, through which the message previously happened to pass, could be broken.

### message

First, has a unique instance ID. This is to check whether the message came from its own instance or not. Then, `namespace`, `instanceId`, and `message type` are checked strictly in the req/res listeners.

### messaging-demo

- Add CustomEventMessage demo
- Add multiple bi-directional messaging demo

#### About vitest `browser mode`

I was annoyed by the flakey test in `page.test`, so I partially switched to `browser mode`. I understand that the `window`(jsdom global context) on nodeVM is common between tests, but can be made isolated when in `browser mode`.

I really wanted to split the test environment nicely with workspace configs [like in this example](https://github.com/vitest-dev/vitest/discussions/5461), but I gave up because of a bug that broke the existing unit test. The bug that cjs mocking (`webextension-polyfill`) fails when using browser mode together. They seem to have reworked the whole mock mechanism. 

This Issue (#48) may have been resolved. But the link is already dead, I don't know which test Flakey was testing at the time.

## Diagrams

> [mermaid live editor link](https://mermaid.live/edit#pako:eNqVlG9v2jAQxr-KZamvGv7YyYBYVSTGWgmpwAs6VZqQJpMc1BOxM8fZoIjvPieBACUVLC8SO3ru7vc4l9viUEWAGU7hdwYyhG-CLzWPZxLZa67WaKCkAWka01CLxDzMdSsobkKmhtsAUuy2SPIY0oSHwNBCKbQrMyRcGxGKhEuDbAT5qSTiKVJyBGnKl_CJKgUZ5br8eaa0-yPaUP6C0EDUeoX5VBi4hKP_A0dvgqM3wvWlMm-gi-Nb15ybe4k25_oTNPcmNPcKWrm8uzvWRF_Ld2NlAKk_oA_FnCodQ4PJaPR9PBz0X4aTMXp6nrwW6NP-6BH1p-j58enlkNsCoXKVX9WnbATB_f6AWcHWMpsE2AwnQi7JDNeFNNb7fmFILKXSlu-vLILP1HnOUkv3uCdqDemlOAgaVRWWS2phqjoXluipJVJriX6wRE8s0auWSGWJXLdEjpZovSV6ZqlcHjtBoVDFcSZFyG0LrIQ85D9rCfqhJcbnYUIWHRGJxQK0HRbHBrPJsINj0DEXkR0z2zz5DNs_I4YZZnYZwYJnK5ND7qyUZ0ZNNzLEzOgMHKxVtnzDbMFXqd1lSWTr7WdU9db2P2ZbvMasQYjX9AnteaRN3HbP7XgO3mBG3U6z533p-W637bue5-4c_K6UTUGbpOtT36edLvXaLu06GCJhlB6Vc7EYj0WJH4U-p9r9Aynbsqc)

Here is the understanding I have in my head and its diagram.

```mermaid
sequenceDiagram
    box Content-Script<br/><br/>instance1<br/>{ namespace: foo }
    participant ins1_on as onMessage
    participant ins1_send as sendMessage
    end
    box Injected/WebSite<br/><br/>instance2<br/>{ namespace: foo }
    participant ins2_on as onMessage
    participant ins2_send as sendMessage
    end
    box AnotherContext<br/><br/>instance3<br/>{ namespace: bar }
    participant ins3_on as onMessage
    participant ins3_send as sendMessage
    end

    %% namespace B
    Note over ins3_on,ins3_send: COMMUNICATION FLOW<br/>SAME AS LEFT

    par 
        ins1_send->>+ins2_on: send/type:"ping1"
        ins1_send--xins1_on: ignore own send
        ins2_on--xins2_send: ignore own res
        ins2_on->>-ins1_send: res/type:"ping1"
    end
    par 
        ins2_send->>+ins1_on: send/type:"ping2"
        ins2_send--xins2_on: ignore own send
        ins1_on--xins1_send: ignore own res
        ins1_on->>-ins2_send: res/type:"ping2"
    end

    
    %% no communicate lines
    Note over ins2_on,ins3_send: No communicate in<br/>different namespaces
```

### CODE REVIEW

File: packages/messaging/src/__tests__/browser/page.test.ts
Comment (1natsu172) [inline packages/messaging/src/__tests__/browser/page.test.ts:1]: Switch `page.test` to browser mode test.

File: packages/messaging/src/custom-event.ts [hunk 1]
```diff
@@ -51,48 +51,63 @@ export type CustomEventMessenger<TProtocolMap extends Record<string, any>> = Gen
...
  * })
+ *
+ * * @link Spec diagrams
+ * https://mermaid.live/edit#pako:eNqVlG9v2jAQxr-KZamvGv7YyYBYVSTGWgmpwAs6VZqQJpMc1BOxM8fZoIjvPieBACUVLC8SO3ru7vc4l9viUEWAGU7hdwYyhG-CLzWPZxLZa67WaKCkAWka01CLxDzMdSsobkKmhtsAUuy2SPIY0oSHwNBCKbQrMyRcGxGKhEuDbAT5qSTiKVJyBGnKl_CJKgUZ5br8eaa0-yPaUP6C0EDUeoX5VBi4hKP_A0dvgqM3wvWlMm-gi-Nb15ybe4k25_oTNPcmNPcKWrm8uzvWRF_Ld2NlAKk_oA_FnCodQ4PJaPR9PBz0X4aTMXp6nrwW6NP-6BH1p-j58enlkNsCoXKVX9WnbATB_f6AWcHWMpsE2AwnQi7JDNeFNNb7fmFILKXSlu-vLILP1HnOUkv3uCdqDemlOAgaVRWWS2phqjoXluipJVJriX6wRE8s0auWSGWJXLdEjpZovSV6ZqlcHjtBoVDFcSZFyG0LrIQ85D9rCfqhJcbnYUIWHRGJxQK0HRbHBrPJsINj0DEXkR0z2zz5DNs_I4YZZnYZwYJnK5ND7qyUZ0ZNNzLEzOgMHKxVtnzDbMFXqd1lSWTr7WdU9db2P2ZbvMasQYjX9AnteaRN3HbP7XgO3mBG3U6z533p-W637bue5-4c_K6UTUGbpOtT36edLvXaLu06GCJhlB6Vc7EYj0WJH4U-p9r9Aynbsqc
  */
 export function defineCustomEventMessaging<
   TProtocolMap extends Record<string, any> = Record<string, any>,
 >(config: CustomEventMessagingConfig): CustomEventMessenger<TProtocolMap> {
   const namespace = config.namespace;
+  const instanceId = crypto.randomUUID();
```
Comment (1natsu172) [inline packages/messaging/src/custom-event.ts:62]: Is there a case where it's not a secure environment? crypto can only be used in a secure environment, if so, we need to use the UUID library.

File: packages/messaging/src/custom-event.ts [hunk 2]
```diff
@@ -51,48 +51,63 @@ export type CustomEventMessenger<TProtocolMap extends Record<string, any>> = Gen
...
       const responseListener = (e: Event) => {
         const { detail } = e as CustomEvent;
-        res(detail);
+        if (
+          detail.namespace === namespace &&
+          detail.instanceId !== instanceId &&
+          detail.message.type === requestEvent.detail.message.type
+        ) {
+          res(detail.response);
+        }
```
Comment (1natsu172) [inline packages/messaging/src/custom-event.ts:76]: Is it a mistake that unlike windowMessage, customEventMessage has been unguarded so far? Is there a particular reason I don't know about?

File: packages/messaging/src/custom-event.ts [hunk 3]
```diff
@@ -51,48 +51,63 @@ export type CustomEventMessenger<TProtocolMap extends Record<string, any>> = Gen
...
       };
       removeAdditionalListeners.push(() =>
         window.removeEventListener(RESPONSE_EVENT, responseListener),
       );
-      window.addEventListener(RESPONSE_EVENT, responseListener, { once: true });
```
Comment (1natsu172) [inline packages/messaging/src/custom-event.ts:70]: Here, too, I don't understand why it has been `once: true` so far.

**Comment by 1natsu172** [comment]
@aklinker1 PR is ready for review. But there's really no need hurry. Because at least in my project, I'm resuming work on this content with a local patching. 😃

File: packages/messaging/src/custom-event.ts [hunk 4]
```diff
@@ -51,48 +52,67 @@ export type CustomEventMessenger<TProtocolMap extends Record<string, any>> = Gen
...
   const messenger = defineGenericMessanging<TProtocolMap, CustomEventMessage, []>({
     ...config,
 
     sendMessage(message) {
-      const event = new CustomEvent(REQUEST_EVENT, { detail: { message, namespace } });
-      return sendCustomMessage(event);
+      const reqDetail = { message, namespace, instanceId };
+      const requestEvent = new CustomEvent(REQUEST_EVENT, {
+        // @ts-expect-error not exist cloneInto types because implemented only in Firefox.
+        detail: typeof cloneInto !== 'undefined' ? cloneInto(reqDetail, window) : reqDetail,
```
Comment (1natsu172) [inline packages/messaging/src/custom-event.ts:93]: There is a problem with `CustomEvent` in Firefox that causes an error when sharing a message object.

image

- https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Sharing_objects_with_page_scripts#sharing_content_script_objects_with_page_scripts
- https://stackoverflow.com/questions/18744224/triggering-a-custom-event-with-attributes-from-a-firefox-extension

**Comment by molvqingtai** [comment]
Waiting for merge....

File: packages/messaging/src/custom-event.ts (hunk 2 above)
Comment (aklinker1) [reply packages/messaging/src/custom-event.ts:77]: Probably a bug, but could be related to how `window.postMessage` sends the message to all frames in the window.

**Review by aklinker1 — APPROVED** [review]
Looks good to me, sorry it took a while to get to! 

I will do a major release for this change.

**Comment by 1natsu172** [comment]
No problem, thanks for the hard review!

**Comment by aklinker1** [comment]
Released [v2.0.0](https://github.com/aklinker1/webext-core/releases/tag/messaging-v2.0.0)

**Comment by 1natsu172** [comment]
@aklinker1 It seems that the published `2.0.0` contents are still in `1.4.0`. There may be a problem with the build process 🤔 .
https://www.npmjs.com/package/@webext-core/messaging?activeTab=code

**Comment by aklinker1** [comment]
@1natsu172 fixed, https://github.com/aklinker1/webext-core/releases/tag/messaging-v2.0.1

**Comment by 1natsu172** [comment]
Thanks, confirmed that the correct update came 🚀

**Pull request merged**